Support Peek & Pop

filletofish opened the pull request:
**Тема ревью**
Добавил peek & pop для `MainViewController` и `PastEventsViewController`

**Описание ревью**
 - Реализация `UIViewControllerPreviewingDelegate`
 - Изменения в соответствующих DisplayCollection

**На что обратить внимание и как тестировать**
Вынесен общий код в протоколы. 

В начале хотел создать наследник протокола UIPreviewingDelegate с полем PreviewingContentProvider, в котором добавить дефолтную имплементацию методов  `previewingContext(_ previewingContext: , commit:)`  и `previewingContext(_ : , viewControllerForLocation:) -> UIViewController?`, чтобы полностью вынести общий код. Но пока компилятор не позволяет так делать (objc функции нельзя в extension для протоколов). Так что полностью вынести общий код не удалось

Тестировать можно на симуляторе, если есть трекпад с Fource Touch. 

**Review**
Add peek & pop for `MainViewController` and `PastEventsViewController`

**Review description**
 - Сonform to UIViewControllerPreviewingDelegate
 - Make changes in corresponding DisplayCollections

**How to test? What pay attention to?**

The feature can be tested with simulators on computers with Force TrackPad.
![previewing](https://user-images.githubusercontent.com/14925971/32689927-655457e0-c6fe-11e7-8c0a-320ce42bfd72.gif)

filletofish committed:
Implement previewing viewcontroller feature

For MainViewController and PastEventsViewController

filletofish committed:
Extract register for previewing in protocol extension

Refactor code

Kirillzzy commented:
@filletofish не мог бы ты добавить к описанию PR картинку, показывающую результат, пожалуйста)

Kirillzzy commented:
А вообще все круто, и код хороший)

Kirillzzy commented on CHMeetupApp/Sources/ViewControllers/Feed/Main/MainViewDisplayCollection.swift:
```diff
@@ -106,6 +106,24 @@ class MainViewDisplayCollection: DisplayCollection, DisplayCollectionAction {
     }
   }
+
+  func preview(at indexPath: IndexPath) -> UIViewController? {
```
Насчет вынести в общий протокол DisplayCollection.
Давай так и сделаем, но чтобы по дефолту его можно было не реализовывать. Например в этом методе возвращать nil в extension реализации.

filletofish replied:
@Kirillzzy  Оки, подумаю в этом направлении

filletofish committed:
Merge develop into improvement/peek-pop-feature

filletofish committed:
Introduce PreviewingContentProvider

Conform main display collection and past events display collection to
PreviewingContentProvider.

filletofish committed:
Refactor common code to protocol extension

filletofish committed:
Merge conflicts

filletofish commented:
@Kirillzzy поправил, добавил гифку, исправил описание.

smejl commented on CHMeetupApp/Sources/Common/Helpers/Previewing/Previewing.swift:
```diff
@@ -0,0 +1,44 @@
+import UIKit
+
+extension UIViewControllerPreviewingDelegate where Self: UIViewController {
+  func registerForPreviewingIfAvailable() {
```
Предлагаю чуть изменить этот метод. Сейчас в названии есть состояние (если доступно), а при этом доступно или нет – мы не знаем. Предлагаю заменить на 
`fund registerForPreviewing() -> Bool`

smejl replied:
И можно добавить второй вариант без возвращаемого значения.

filletofish replied:
👍

filletofish replied:
@smejl done

filletofish committed:
Refactor by @smejl pr comments

filletofish committed:
Merge branch 'cocoaheadsru/develop' into develop

filletofish committed:
Merge branch 'develop' into improvement/peek-pop-feature

Kirillzzy commented on CHMeetupApp/Sources/Common/Helpers/Previewing/Previewing.swift:
```diff
@@ -0,0 +1,48 @@
+
+extension UIViewControllerPreviewingDelegate where Self: UIViewController {
+  @discardableResult
+  func registerForPreviewing() -> Bool {
```
Только лучше ‘registeredForViewController’. Иначе обозначает процесс, а он не может ничего возвращать. Исправь это и апрув

filletofish replied:
Он и должен обозначать процесс, потому что это тут главное. Лучше уж убрать возвращаемое значение: не могу представить кэйс, где оно могло бы пригодиться.
